lnrpc/routerrpc/router_backend.go

```diff
@@ -353,22 +409,144 @@ func (r *RouterBackend) parseQueryRoutesRequest(in *lnrpc.QueryRoutesRequest) (
+	return routing.NewRouteRequest(
+		sourcePubKey, targetPubKey, amt, in.TimePref, restrictions,
+		customRecords, routeHintEdges, blindedPmt, finalCLTVDelta,
+	)
+}
+
+func parseBlindedPayment(in *lnrpc.QueryRoutesRequest) (
+	*routing.BlindedPayment, error) {
+
+	if len(in.PubKey) != 0 {
+		return nil, fmt.Errorf("target pubkey: %x should not be set "+
+			"when blinded path is provided", in.PubKey)
+	}
+
+	if len(in.BlindedPaymentPaths) != 1 {
+		return nil, errors.New("query routes only supports a single " +
+			"blinded path")
+	}
+
+	blindedPath := in.BlindedPaymentPaths[0]
+
+	if len(in.RouteHints) > 0 {
+		return nil, errors.New("route hints and blinded path can't " +
+			"both be set")
+	}
+
+	blindedPmt, err := unmarshalBlindedPayment(blindedPath)
+	if err != nil {
+		return nil, fmt.Errorf("parse blinded payment: %w", err)
+	}
+
+	if err := blindedPmt.Validate(); err != nil {
+		return nil, fmt.Errorf("invalid blinded path: %w", err)
+	}
+
+	if in.FinalCltvDelta != 0 {
+		return nil, errors.New("final cltv delta should be " +
+			"zero for blinded paths")
+	}
+
+	// For blinded paths, we get one set of features for the relaying
+	// intermediate nodes and the final destination. We don't allow the
+	// destination feature bit field for regular payments to be set, as
+	// this could lead to ambiguity.
+	if len(in.DestFeatures) > 0 {
+		return nil, errors.New("destination features should " +
+			"be populated in blinded path")
+	}
+
+	return blindedPmt, nil
+}
+
+func unmarshalBlindedPayment(rpcPayment *lnrpc.BlindedPaymentPath) (
+	*routing.BlindedPayment, error) {
+
+	if rpcPayment == nil {
+		return nil, errors.New("nil blinded payment")
+	}
+
+	path, err := unmarshalBlindedPaymentPaths(rpcPayment.BlindedPath)
 	if err != nil {
 		return nil, err
 	}
-	routeHintEdges, err := routing.RouteHintsToEdges(
-		routeHints, targetPubKey,
-	)
+
+	features, err := UnmarshalFeatures(rpcPayment.Features)
 	if err != nil {
 		return nil, err
 	}
 
-	return routing.NewRouteRequest(
-		sourcePubKey, &targetPubKey, amt, in.TimePref, restrictions,
-		customRecords, routeHintEdges, nil, finalCLTVDelta,
-	)
+	return &routing.BlindedPayment{
+		BlindedPath:     path,
+		CltvExpiryDelta: uint16(rpcPayment.TotalCltvDelta),
+		BaseFee:         uint32(rpcPayment.BaseFeeMsat),
+		ProportionalFee: uint32(
+			rpcPayment.ProportionalFeeMsat,
+		),
+		HtlcMinimum: rpcPayment.HtlcMinMsat,
+		HtlcMaximum: rpcPayment.HtlcMaxMsat,
+		Features:    features,
+	}, nil
+}
+
+func unmarshalBlindedPaymentPaths(rpcPath *lnrpc.BlindedPath) (
+	*sphinx.BlindedPath, error) {
+
+	if rpcPath == nil {
+		return nil, errors.New("blinded path required when blinded " +
+			"route is provided")
+	}
+
+	introduction, err := btcec.ParsePubKey(rpcPath.IntroductionNode)
+	if err != nil {
+		return nil, err
+	}
+
+	blinding, err := btcec.ParsePubKey(rpcPath.BlindingPoint)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(rpcPath.BlindedHops) < 1 { //nolint:gomnd
```

Comment:
think we've disabled `gomnd` after a few complains😂

Reply:
😈